kornia/feature/sold2/sold2.py

```diff
@@ -0,0 +1,368 @@
+
+    def filter_and_match_lines(self, scores):
+        """Use the scores to keep the top k best lines, compute the Needleman- Wunsch algorithm on each candidate
+        pairs, and keep the highest score.
+
+        Inputs:
+            scores: a (N, M, n, n) np.array containing the pairwise scores
+                    of the elements to match.
+        Outputs:
+            matches: a (N) np.array containing the indices of the best match
+        """
+        # Pre-filter the pairs and keep the top k best candidate lines
+        line_scores1 = scores.max(3)
+        valid_scores1 = line_scores1 != -1
+        line_scores1 = ((line_scores1 * valid_scores1).sum(2)
+                        / valid_scores1.sum(2))
+        line_scores2 = scores.max(2)
+        valid_scores2 = line_scores2 != -1
+        line_scores2 = ((line_scores2 * valid_scores2).sum(2)
+                        / valid_scores2.sum(2))
+        line_scores = (line_scores1 + line_scores2) / 2
+        topk_lines = np.argsort(line_scores,
+                                axis=1)[:, -self.top_k_candidates:]
+        # topk_lines.shape = (n_lines1, top_k_candidates)
+        top_scores = np.take_along_axis(scores, topk_lines[:, :, None, None],
+                                        axis=1)
+
+        # Consider the reversed line segments as well
+        top_scores = np.concatenate([top_scores, top_scores[..., ::-1]],
+                                    axis=1)
+
+        # Compute the line distance matrix with Needleman-Wunsch algo and
+        # retrieve the closest line neighbor
+        n_lines1, top2k, n, m = top_scores.shape
+        top_scores = top_scores.reshape((n_lines1 * top2k, n, m))
+        nw_scores = self.needleman_wunsch(top_scores)
+        nw_scores = nw_scores.reshape(n_lines1, top2k)
+        matches = np.mod(np.argmax(nw_scores, axis=1), top2k // 2)
+        matches = topk_lines[np.arange(n_lines1), matches]
+        return matches
+
+    def needleman_wunsch(self, scores):
+        """Batched implementation of the Needleman-Wunsch algorithm.
+
+        The cost of the InDel operation is set to 0 by subtracting the gap
+        penalty to the scores.
+        Inputs:
+            scores: a (B, N, M) np.array containing the pairwise scores
+                    of the elements to match.
+        """
+        b, n, m = scores.shape
+
+        # Recalibrate the scores to get a gap score of 0
+        gap = 0.1
+        nw_scores = scores - gap
+
+        # Run the dynamic programming algorithm
+        nw_grid = np.zeros((b, n + 1, m + 1), dtype=float)
+        for i in range(n):
+            for j in range(m):
+                nw_grid[:, i + 1, j + 1] = np.maximum(
+                    np.maximum(nw_grid[:, i + 1, j], nw_grid[:, i, j + 1]),
+                    nw_grid[:, i, j] + nw_scores[:, i, j])
+
+        return nw_grid[:, -1, -1]
+
+
+def keypoints_to_grid(keypoints: torch.Tensor, img_size: tuple) -> torch.Tensor:
```

Comment:
we have https://kornia.readthedocs.io/en/latest/geometry.conversions.html#kornia.geometry.conversions.normalize_pixel_coordinates